### ISSUE TITLE
Encoding issue in headers data

### ISSUE THREAD

**Issue by SharonAz** [renatoathaydes/rawhttp#28]
When headers data contains UTF8 characters, the value data returns an unreadable string
I have an HTTP request with UTF8 headers values
for example:
POST /app/v3.1.0/sample-api HTTP/1.1
HOST: api.example.com
Application-Id: B30F8BB0-3B37-11E7-8C40-000D3A196736
Connection: keep-alive
X-TEST: こんにちは
Content-Type: application/json

I parsed the request data using RawHttp.
When I read the header 'X-TEST' value, I get garbage string instead of the Japanese string

Code example:
RawHttp rawHttp = new RawHttp();
rawHttp.parseRequest(_requestData);
String headerValue = _httpRequest.getHeaders().get("X-TEST");

I also tried to provide InputStream to rawHttp.parseRequest and I get the same result.
I could not find any option to parse the headers as UTF8.

Thanks,
Sharon

**Comment by renatoathaydes** [comment]
HTTP Headers are encoded in US-ASCII, but allows any bytes in the field-value for historical reasons.

From the RFC:

```
field-value    = *( field-content / obs-fold )
     field-content  = field-vchar [ 1*( SP / HTAB ) field-vchar ]
     field-vchar    = VCHAR / obs-text
```

and:

```
VCHAR (any
   visible [USASCII] character).
```

and:

```
obs-text       = %x80-FF
```

However, https://tools.ietf.org/html/rfc7230#section-3.2.4:

```
 Historically, HTTP has allowed field content with text in the
   ISO-8859-1 charset [ISO-8859-1], supporting other charsets only
   through use of [RFC2047] encoding.  In practice, most HTTP header
   field values use only a subset of the US-ASCII charset [USASCII].
   Newly defined header fields SHOULD limit their field values to
   US-ASCII octets.  A recipient SHOULD treat other octets in field
   content (obs-text) as opaque data.
```

RawHTTP seems to correctly follow the spec, I think. There's no way we can use UTF-8 to decode header values.

It may be possible to add an option to retrieve header values as bytes, which could then preserve the unicode which would allow you to decode a value as UTF-8.

I will look into it.

**Comment by renatoathaydes** [comment]
I am currently working on this.

**Comment by renatoathaydes** [comment]
Hey @SharonAz this Kotlin sample shows how you will be able to use HTTP Header values (the names are still restricted by the spec) with whatever encoding you want: https://github.com/renatoathaydes/rawhttp/blob/e352871d9a31d7bdebbedc29308d82449a9e9d22/samples/src/test/kotlin/rawhttp/samples/server-and-client-rawhttp.kt

**Comment by renatoathaydes** [comment]
Released in 2.4.0.
Added a note in the docs about how to encode header values using UTF: https://renatoathaydes.github.io/rawhttp/docs/low-level-api.html
